core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -136,7 +139,21 @@ private[spark] class TaskSchedulerImpl(
   // IDs of the tasks running on each executor
   private val executorIdToRunningTaskIds = new HashMap[String, HashSet[Long]]
 
-  private val executorsPendingDecommission = new HashMap[String, ExecutorDecommissionInfo]
+  // We add executors here when we first get decommission notification for them. Executors can
+  // continue to run even after being asked to decommission, but they will eventually exit.
+  val executorsPendingDecommission = new HashMap[String, ExecutorDecommissionInfo]
+
+  // When they exit and we know of that via heartbeat failure, we will add them to this cache.
+  // This cache is consulted to know if a fetch failure is because a source executor was
+  // decommissioned.
+  lazy val decommissionedExecutorsRemoved = CacheBuilder.newBuilder()
+    .expireAfterWrite(
+      conf.getLong("spark.decommissioningRememberAfterRemoval.seconds", 60L), TimeUnit.SECONDS)
```

Comment:
Not related to this PR but just for double-check: The fetch failure may arrive after the executor is marked as removed, even without the "early exit" feature, right? For example, the RPC framework has a huge delay. So this is a best-effort anyway.

Reply:
Yeah the fetch failure is most likely going to arrive well after the executor is removed because the reducer typically retries before giving up with a fetch failure. 

The idea here is to be best effort: to act on fetch failures stemming from decommissioned mappers and purge out all of the other mappers on that host asap. Ofcourse, it is possible that the fetch failure arrives too late and we have forgotten that this was a decommissioned host. 

We are not trying to promise 100% robustness with decommissioning, just trying to improve it a bit. 

So now one question please: what should be the value of this timeout. I am thinking more like 5 minutes ? Obviously the smaller it is, the less the space overhead. But I think 5 minutes should be fine. how much of a delay have you seen in practice ? What is the delay in the RPC framework we are talking about ?

Reply:
I'm not sure how frequent decommission happens. Each entry here should take less than 1kb, so even decommission happens once per second, it's less than 300kb.

Reply:
BTW this is core and we can define the config in `org.apache.spark.internal.config`.

According to other decommission related configs, how about `spark.driver.decommission.infoCacheTTLInSeconds`?